encoding/delta/binary_packed_amd64.s

```diff
@@ -0,0 +1,819 @@
+    VPSUBD Y0, Y1, Y1                   \
+    VPSUBD Y0, Y2, Y2                   \
+    VPSUBD Y0, Y3, Y3                   \
+    VPSUBD Y0, Y4, Y4                   \
+    VMOVDQU Y1, offset+0(baseAddr)      \
+    VMOVDQU Y2, offset+32(baseAddr)     \
+    VMOVDQU Y3, offset+64(baseAddr)     \
+    VMOVDQU Y4, offset+96(baseAddr)
+
+// func blockSubInt32AVX2(block *[blockSize]int32, value int32)
+TEXT ·blockSubInt32AVX2(SB), NOSPLIT, $0-16
+    MOVQ block+0(FP), AX
+    VPBROADCASTD value+8(FP), Y0
+    subInt32AVX2x32(AX, 0)
+    subInt32AVX2x32(AX, 128)
+    subInt32AVX2x32(AX, 256)
+    subInt32AVX2x32(AX, 384)
+    VZEROUPPER
+    RET
+
+// func blockBitWidthsInt32AVX2(bitWidths *[numMiniBlocks]byte, block *[blockSize]int32)
+TEXT ·blockBitWidthsInt32AVX2(SB), NOSPLIT, $0-16
+    MOVQ bitWidths+0(FP), AX
+    MOVQ block+8(FP), BX
+
+    // AVX2 only has signed comparisons (and min/max), we emulate working on
+    // unsigned values by adding -2^31 to the values. Y5 is a vector of -2^31
+    // used to offset 8 packed 32 bits integers in other YMM registers where
+    // the block data are loaded.
+    VPCMPEQD Y5, Y5, Y5
+    VPSLLD $31, Y5, Y5
+
+    XORQ DI, DI
+loop:
+    VPBROADCASTD (BX), Y0 // max
+    VPADDD Y5, Y0, Y0
+
+    VMOVDQU (BX), Y1
+    VMOVDQU 32(BX), Y2
+    VMOVDQU 64(BX), Y3
+    VMOVDQU 96(BX), Y4
+
+    VPADDD Y5, Y1, Y1
+    VPADDD Y5, Y2, Y2
+    VPADDD Y5, Y3, Y3
+    VPADDD Y5, Y4, Y4
+
+    VPMAXSD Y2, Y1, Y1
+    VPMAXSD Y4, Y3, Y3
+    VPMAXSD Y3, Y1, Y1
+    VPMAXSD Y1, Y0, Y0
+
+    VPERM2I128 $1, Y0, Y0, Y1
+    VPMAXSD Y1, Y0, Y0
+
+    VPSHUFD $0b00011011, Y0, Y1
+    VPMAXSD Y1, Y0, Y0
+    VPSUBD Y5, Y0, Y0
+
+    MOVQ X0, CX
+    MOVL CX, DX
+    SHRQ $32, CX
+    CMPL CX, DX
+    CMOVLHI CX, DX
+
+    LZCNTL DX, DX
+    NEGL DX
+    ADDL $32, DX
+    MOVB DX, (AX)(DI*1)
+
+    ADDQ $128, BX
+    INCQ DI
+    CMPQ DI, $numMiniBlocks
+    JNE loop
+    VZEROUPPER
+    RET
+
+// miniBlockPackInt32Default is the generic implementation of the algorithm to
+// pack 32 bit integers into values of a given bit width (<=32).
+//
+// This algorithm is much slower than the vectorized versions, but is useful
+// as a reference implementation to run the tests against, and as fallback when
+// the code runs on a CPU which does not support the AVX2 instruction set.
+//
+// func miniBlockPackInt32Default(dst *byte, src *[miniBlockSize]int32, bitWidth uint)
+TEXT ·miniBlockPackInt32Default(SB), NOSPLIT, $0-24
+    MOVQ dst+0(FP), AX
+    MOVQ src+8(FP), BX
+    MOVQ bitWidth+16(FP), R9
+
+    XORQ DI, DI // bitOffset
+    XORQ SI, SI
+loop:
+    MOVQ DI, CX
+    MOVQ DI, DX
+
+    ANDQ $0b11111, CX // bitOffset % 32
+    SHRQ $5, DX       // bitOffset / 32
+
+    MOVLQZX (BX)(SI*4), R8
+    SHLQ CX, R8
+    ORQ R8, (AX)(DX*4)
+
+    ADDQ R9, DI
+    INCQ SI
+    CMPQ SI, $miniBlockSize
+    JNE loop
+    RET
+
+// miniBlockPackInt32x1bitAVX2 packs 32 bit integers into 1 bit values in the
```

Comment:
I'm confused about how you can pack a 32 bit integer into a 1 bit value.

Reply:
The encoding first computes the bit width of the largest delta. If that value is 0 or 1 the minimum bit width needed to represent the values is 1 since all upper bits will be zero then. In that case we can pack the 32 bits integers into single bit values.